### PR TITLE
Refactor to use real constructors in factory

### PR DESCRIPTION
**Pull request by bitwiseman**
Removed need for `init` method, allowing us to use `final` on fields that should not change. 

Also fixed missing assignment to name field in base class.

### CODE REVIEW

File: src/main/java/jenkins/plugins/logstash/persistence/AbstractLogstashIndexerDao.java [hunk 1]
```diff
@@ -37,16 +37,17 @@
...
+  protected final String key;
+  protected final String username;
+  protected final String password;
 
-  void init(String host, int port, String key, String username, String password) {
+  AbstractLogstashIndexerDao(String host, int port, String key, String username, String password) {
     this.host = host;
     this.port = port;
     this.key = key;
+    this.username = username;
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/persistence/AbstractLogstashIndexerDao.java]: Don't know why this wasn't here to begin with.

File: src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java [hunk 2]
```diff
@@ -80,12 +79,13 @@ final void init(String host, int port, String key, String username, String passw
...
 
-    auth = null;
     if (StringUtils.isNotBlank(username)) {
       auth = Base64.encodeBase64String((username + ":" + StringUtils.defaultString(password)).getBytes());
+    } else {
+      auth = null;
     }
 
-    clientBuilder = HttpClientBuilder.create();
+    clientBuilder = factory == null ? HttpClientBuilder.create() : factory;
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/persistence/ElasticSearchDao.java]: This allows us to pass this field in for mocking or just initialize normally.  This makes for the least difference between mocking and normal behavior.

File: src/main/java/jenkins/plugins/logstash/persistence/IndexerDaoFactory.java [hunk 3]
```diff
@@ -73,20 +74,24 @@ public static synchronized LogstashIndexerDao getInstance(IndexerType type, Stri
       throw new InstantiationException("Unknown IndexerType '" + type + "'.");
     }
 
-    if (shouldRefreshInstance(type, host, port, key, password)) {
-      try {
-        instance = (AbstractLogstashIndexerDao) INDEXER_MAP.get(type).newInstance();
+    if (shouldRefreshInstance(type, host, port, key, username, password)) {
+    try {
+        instance = (AbstractLogstashIndexerDao) INDEXER_MAP.get(type).getConstructor(String.class, int.class, String.class, String.class, String.class)
```
Comment (bitwiseman) [inline src/main/java/jenkins/plugins/logstash/persistence/IndexerDaoFactory.java]: No reason we can't select and call a constructor with parameters.

**Comment by R-Gerard** [comment]
Looks good. Can you fix the whitespace in IndexerDaoFactory.java? The indentation in the try/catch block is a little wonky.

**Comment by bitwiseman** [comment]
Done.

**Comment by bitwiseman** [comment]
Done for real.

**Pull request merged**